orangecontrib/text/preprocess.py

```diff
@@ -1,94 +1,161 @@
-        :param stop_words: Determines whether stop words should("english"), or should not(None) be removed.
-            If this is list, it should contain stopwords.
-        :type stop_words: 'english' or list or None
-        :param trans: An optional pre-processor object to perform the morphological
-            transformation on the tokens before returning them.
-        :type trans: :class: `orangecontrib.text.preprocess.Lemmatizer`
-            or :class: `orangecontrib.text.preprocess.Stemmer`
-        :return: :class: `orangecontrib.text.preprocess.Preprocessor`
+        :param stop_words: Determines whether and what stop words should be removed. Can remove
+            default nltk stop words for the English language or stop words provided in a custom list.
+        :type stop_words: 'default', a list or None
+        :param trans: Name of the morphological transformation method to be performed on the tokens.
+        :type trans: 'porter_stemmer', 'snowball_stemmer', 'lemmatizer' or None
+        :return: list
         """
-        # TODO Needs more elaborate check on list contents.
-        if not (stop_words == 'english' or isinstance(stop_words, list) or stop_words is None):
-            raise ValueError("The stop words parameter should be either \'english\', a list or None.")
-        if isinstance(trans, Stemmatizer):  # hack, since scikit does not obey lowercase when preprocesor given
-            trans.lowercase = lowercase
-        self.stop_words = stop_words
-        # TODO does including punctuation even work?
-        self.incl_punct = incl_punct
+        # Tokenizer.
+        if tokenizer is None:
+            raise ValueError("The pre-processor must specify a tokenizer "
+                             "('word_tokenizer' or 'twitter_tokenizer').")
+        elif tokenizer == "word_tokenizer":
+            self.tokenizer = word_tokenize
+        elif tokenizer == "twitter_tokenizer":
+            self.tokenizer = word_tokenize
+        else:
+            raise ValueError("Tokenizer of type '{}', is not recognized.".format(tokenizer))
+
+        # Lowercase.
         self.lowercase = lowercase
-        self.transformation = trans
-        self.min_df = min_df
-        self.cv = CountVectorizer(
-            lowercase=lowercase,
-            stop_words=stop_words,
-            preprocessor=trans,
-            min_df=min_df,
-        )
+
+        # Stop words.
+        if stop_words == "default":
+            self.stop_words = stopwords.words("english")
+        elif isinstance(stop_words, list):
+            self.stop_words = stop_words
+        elif stop_words is None:
+            self.stop_words = None
+        else:
+            raise ValueError("The stop words parameter should be either \'default\', "
+                             "a list containing the stop words or None.")
+
+        # Transformation.
+        self.transformation = transformation
+        if transformation == 'porter_stemmer':
```

Comment:
Why don't we use the objects of type `Stemmatizer` directly? Then, all these following line cound be removed and we also remove the unnecessary mapping of these names to objects.